src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -1454,6 +1454,102 @@ void MacroAssembler::update_word_crc32(Register crc, Register v, Register tmp1,
   xorr(crc, crc, tmp2);
 }
 
+
+// This improvement (vectorization) is based on java.base/share/native/libzip/zlib/zcrc32.c.
+// To make it, following steps are taken:
+//  1. in zcrc32.c, modify N to 16 and related code,
+//  2. re-generate the tables needed, we use tables of (N == 16, W == 4)
+//  3. finally vectorize the code (original implementation in zcrc32.c is just scalar code).
+// New tables for vector version is after table3.
+void MacroAssembler::vector_update_crc32(Register crc, Register buf, Register len,
+                                         Register tmp1, Register tmp2, Register tmp3, Register tmp4,
+                                         Register table0, Register table3) {
+    const int N = 16, W = 4;
+    const int64_t single_table_size = 256;
+    const Register blks = tmp2;
+    const Register tmpTable = tmp3, tableN16 = tmp4;
+    const VectorRegister vcrc = v4, vword = v8, vtmp = v12;
+    Label VectorLoop;
+    Label LastBlock;
+
+    add(tableN16, table3, 1*single_table_size*sizeof(juint), tmp1);
+    mv(t0, 0xff);
```

Comment:
Thanks for the quick update. But I am a bit worried about keeping a long-lived value in 't0'. `t0` as a scratch register are implicitly used by various assembler routines, so I think it's error-prone to do this. Can you choose another one? Seems `tmp5` or `tmp6`  in `kernel_crc32` are usable for our purpose.

Reply:
Make sense, fixed. Thanks!

Reply:
Ah, seems previously I forgot to push the change.